### Importing required libraries :

In [1]:
import numpy as np
%matplotlib inline
import matplotlib.pyplot as plt
#from random import normal
import math
from sklearn.metrics import accuracy_score

### Creating gaussian distributions : 

In [2]:
def createdata(v):
    
    np.random.seed(0)
    dis1=np.random.normal(0,v,50)
    dis2=np.random.normal(5,v,50)
    dis3=np.random.normal(10,v,50)
    A=np.concatenate([dis1,dis2])
    X=np.concatenate([A,dis3])
    X
    return X


In [3]:
X=createdata(1)
X.shape

(150,)

### Plotting the data:

In [4]:
'''
def plot_hist(data):
    plt.figure(figsize=(15,7))
    for x in data:
        plt.hist(x, bins = 80, normed = True, alpha = 0.7)
    #plt.xlim(-10, 20)
    plt.figure(figsize=(15,7))
    
plot_hist(X)
'''

'\ndef plot_hist(data):\n    plt.figure(figsize=(15,7))\n    for x in data:\n        plt.hist(x, bins = 80, normed = True, alpha = 0.7)\n    #plt.xlim(-10, 20)\n    plt.figure(figsize=(15,7))\n    \nplot_hist(X)\n'

### Function to initialize random means, variances and prior probabilities: 

In [5]:
def initialize(X,K):
    
    m=np.array([2.,3.,7.])             #random means
    #m=np.random.choice(X,size=3)      
    v=np.ones((3))                     #random variances set to 1 initially
    w=np.ones((K))/K                   #random prior weights set to equal
    #posterior probability that each input belongs to each of the 3 gaussians
    posterior=np.zeros((len(X),K),dtype=float)
    return m,v,w,posterior

### True assignment of data samples to respective Gaussians and means:

In [6]:
Y=[0]*50 + [1]*50 + [2]*50
Y
M=[0,5,10]
V=[1,1,1]

### Function to evaluate accuracy:

In [7]:
def accuracy(pred,true,K):
    diff=[]
    summ=0    
    for j in range(K):
        #d=abs(pred[j]-true[j])/true[j]
        diff.append((abs(pred[j]-true[j])))
    summ=sum(diff)
    accu=round(((1-summ)*100),2)    
    #accu=summ
        
    return summ,accu

### Function to allocate the probability of each of the data points belonging to each one of the clusters(Soft Allocation):

In [8]:
def GMM(X,K,epochs):
    N=len(X)
    #K=3
    pi=math.pi
    iterations=epochs
    M=[0.,5.,10.]
    
    for iter in range(iterations):
        
        #Expection Step : Calculating Posterior probabilities for 
        for i in range(len(X)):
            for k in range(K):

                part1=1/math.sqrt(2*pi*(var[k]))
                part2=(-1/2)*((X[i]-means[k])**2)
                part3= part2/(var[k])
                part4=np.exp(part3)
                posterior[i][k]=part1 * part4

                #posterior[i][k]=wprior[k]*(1/math.sqrt(2*var[k]*pi))*(math.exp(-(1/2)*((X[i]-means[k])**2)))

            posterior[i]=posterior[i]/np.sum(posterior[i])
        
        
        #Maximization Step : Updating parameters : mean, variance, and prior probabilities :
        meanspred=[]
        varpred=[]
        diffmean=[]
        lossmean=[]
        for i in range(K):
            wprior[i]=np.sum(posterior[:,i])/N
            means[i]=np.dot(posterior[:,i],X[:].T)/(np.sum(posterior[:,i]))
            var[i]=np.dot(posterior[:,i],((X-means[i])**2).T)/(np.sum(posterior[:,i]))
            meanspred.append(means[i])
            varpred.append(var[i])
        
        ##### Retrieving Predicted class labels from posterior probabilities :
        li=[]
        for i in range(len(X)):
            li.append(np.argmax(posterior[i]))
            
        ######Accuracy Calculation for Mean and Variance at each iteration :
        meanloss,accumean=accuracy(meanspred,M,K)
        varloss,accuvar=accuracy(varpred,V,K)
        ######
        
        
        print("\n")    
        print("Epoch " + str(iter+1) + " : " + "priorprob="+ str(wprior))
        print("Epoch " + str(iter+1) + " : " + "mean="+ str(means))
        print("Epoch " + str(iter+1) + " : "+ "var=" + str(var))
        
        
        #print(meanspred)
        print("Total mean loss is=> ", meanloss)
        print("Accuracy of means after Epoch {} : {}%".format(iter+1, accumean))
        print("Total variance loss is=> ", varloss)
        print("Accuracy of variance after Epoch {} : {}%".format(iter+1, accuvar))
        totalaccuracy = round((accuracy_score(li, Y)*100),2)
        print("Accuracy of allocation after Epoch {} : {}%".format(iter+1, totalaccuracy))
        #totalaccuracy
        
        
    return (wprior, means,var,totalaccuracy,li)
    #means.shape


In [9]:
means,var,wprior,posterior=initialize(X,3)
#means,var,wprior,posterior.shape
wprior,means,var,totalaccuracy,li=GMM(X,3,4)



Epoch 1 : priorprob=[0.31686045 0.18882409 0.49431546]
Epoch 1 : mean=[0.33717436 3.66738246 8.75161024]
Epoch 1 : var=[2.35235318 2.39087316 5.5254408 ]
Total mean loss is=>  2.91818165515591
Accuracy of means after Epoch 1 : -191.82%
Total variance loss is=>  7.268667138969127
Accuracy of variance after Epoch 1 : -626.87%
Accuracy of allocation after Epoch 1 : 82.0%


Epoch 2 : priorprob=[0.29596336 0.28383699 0.42019965]
Epoch 2 : mean=[0.07440818 4.23741447 9.28279348]
Epoch 2 : var=[1.55502693 2.29544849 4.62908539]
Total mean loss is=>  1.5542002307001992
Accuracy of means after Epoch 2 : -55.42%
Total variance loss is=>  5.479560812105378
Accuracy of variance after Epoch 2 : -447.96%
Accuracy of allocation after Epoch 2 : 94.0%


Epoch 3 : priorprob=[0.30453275 0.31415929 0.38130796]
Epoch 3 : mean=[0.01434501 4.55744772 9.67525195]
Epoch 3 : var=[1.18304609 1.63451487 3.33138189]
Total mean loss is=>  0.7816453335035896
Accuracy of means after Epoch 3 : 21.84%
Total variance 

###  Training Accuracy after last epoch :

In [10]:
var

array([1.15981032, 1.13903816, 2.09969777])

In [11]:
finalaccuracy = round((accuracy_score(li, Y)*100),2)
print("Final Accuracy of allocation : {}%".format(finalaccuracy))
print("Number of samples classifed for Gaussians 0, 1, 2 : ",li.count(0),li.count(1),li.count(2))


Final Accuracy of allocation : 98.0%
Number of samples classifed for Gaussians 0, 1, 2 :  48 51 51


* Initially each of the Gaussians consisted of 50 data points. Here we can see that it successfully predicted and classified the total 150 data points in 3 respective Gaussians(0,1,2). 
* Here I ran my algortithm for 4 epochs since I noticed that after 4th epoch my algorithm started to show no change in parameters which are mean, variance and prior probabilities. Also, when I how many data points were part of each cluster, a

### Trying different variances :

### Configuration 2:

In [12]:
X2=createdata(3)
means,var,wprior,posterior=initialize(X2,3)
wprior,means,var,totalaccuracy2,li2=GMM(X2,3,10)
finalaccuracy2 = round((accuracy_score(li2, Y)*100),2)
print("Final Accuracy of allocation : {}%".format(finalaccuracy2))
print("Number of samples classifed for Gaussians 0, 1, 2 : ",li2.count(0),li2.count(1),li2.count(2))



Epoch 1 : priorprob=[0.30379591 0.16129193 0.53491216]
Epoch 1 : mean=[-0.41752964  3.32022977  9.28651244]
Epoch 1 : var=[ 7.26556547  1.90157733 10.5107976 ]
Total mean loss is=>  2.8107874259869305
Accuracy of means after Epoch 1 : -181.08%
Total variance loss is=>  16.677940397926992
Accuracy of variance after Epoch 1 : -1567.79%
Accuracy of allocation after Epoch 1 : 66.0%


Epoch 2 : priorprob=[0.27202643 0.24936382 0.47860975]
Epoch 2 : mean=[-0.52296372  3.70009393  9.60227255]
Epoch 2 : var=[ 8.99009478  2.86158074 11.11711744]
Total mean loss is=>  2.220597241527668
Accuracy of means after Epoch 2 : -122.06%
Total variance loss is=>  19.968792964705695
Accuracy of variance after Epoch 2 : -1896.88%
Accuracy of allocation after Epoch 2 : 72.67%


Epoch 3 : priorprob=[0.27711229 0.28848778 0.43439994]
Epoch 3 : mean=[-0.38087645  4.07237943  9.9145151 ]
Epoch 3 : var=[ 9.60834796  3.74225517 11.1201174 ]
Total mean loss is=>  1.3939819191528322
Accuracy of means after Epoch 3

### Configuration 3 :

In [13]:
X3=createdata(5)
means,var,wprior,posterior=initialize(X3,3)
wprior,means,var,totalaccuracy3,li3=GMM(X3,3,10)
finalaccuracy3 = round((accuracy_score(li3, Y)*100),2)
print("Final Accuracy of allocation : {}%".format(finalaccuracy3))
print("Number of samples classifed for Gaussians 0, 1, 2 : ",li3.count(0),li3.count(1),li3.count(2))



Epoch 1 : priorprob=[0.32099957 0.14138499 0.53761544]
Epoch 1 : mean=[-1.54752871  3.31178119 10.51948588]
Epoch 1 : var=[13.98669165  1.8805     18.07563298]
Total mean loss is=>  3.7552334015839577
Accuracy of means after Epoch 1 : -275.52%
Total variance loss is=>  30.94282463249006
Accuracy of variance after Epoch 1 : -2994.28%
Accuracy of allocation after Epoch 1 : 55.33%


Epoch 2 : priorprob=[0.28882951 0.22971814 0.48145235]
Epoch 2 : mean=[-1.54117813  3.5851093  10.90137209]
Epoch 2 : var=[18.62759833  3.00261374 19.44275524]
Total mean loss is=>  3.8574409197300854
Accuracy of means after Epoch 2 : -285.74%
Total variance loss is=>  38.07296731621261
Accuracy of variance after Epoch 2 : -3707.3%
Accuracy of allocation after Epoch 2 : 57.33%


Epoch 3 : priorprob=[0.29346914 0.26741311 0.43911776]
Epoch 3 : mean=[-1.24711314  3.90304076 11.27074235]
Epoch 3 : var=[21.06948723  4.31769766 19.66699698]
Total mean loss is=>  3.614814729350188
Accuracy of means after Epoch 3 :

### Configuration 4 : 

In [14]:
X4=createdata(10)
means,var,wprior,posterior=initialize(X4,3)
wprior,means,var,totalaccuracy4,li4 = GMM(X4,3,10)
finalaccuracy4 = round((accuracy_score(li4, Y)*100),2)
print("Final Accuracy of allocation : {}%".format(finalaccuracy4))
print("Number of samples classifed for Gaussians 0, 1, 2 : ",li4.count(0),li4.count(1),li4.count(2))



Epoch 1 : priorprob=[0.38434615 0.08659788 0.52905597]
Epoch 1 : mean=[-4.6580071   3.11694109 14.69448081]
Epoch 1 : var=[37.54829617  2.02034978 48.59829648]
Total mean loss is=>  11.235546817487464
Accuracy of means after Epoch 1 : -1023.55%
Total variance loss is=>  85.16694242040813
Accuracy of variance after Epoch 1 : -8416.69%
Accuracy of allocation after Epoch 1 : 43.33%


Epoch 2 : priorprob=[0.3477011  0.17256861 0.47973029]
Epoch 2 : mean=[-4.67859324  3.1957804  15.27753612]
Epoch 2 : var=[48.64916174  3.46394392 51.86050917]
Total mean loss is=>  11.760348961157352
Accuracy of means after Epoch 2 : -1076.03%
Total variance loss is=>  100.97361482676692
Accuracy of variance after Epoch 2 : -9997.36%
Accuracy of allocation after Epoch 2 : 44.0%


Epoch 3 : priorprob=[0.34473224 0.2027475  0.45252027]
Epoch 3 : mean=[-4.41569306  3.30310175 15.703987  ]
Epoch 3 : var=[54.56295141  5.66824926 51.36808334]
Total mean loss is=>  11.81657832163814
Accuracy of means after Epoch 

* We can see a decrease in accuracies when we ran our algorithm for less number of epochs. In our algorithm, we stopped when our posterior probabilities showed almost no change. After trying different values for variance, we noticed that as variance increases, our GMM algorithm takes many iterations to converge. GMM is much more flexible allowing us to generate much better fitting distributions.